File: src/main/java/com/fobgochod/git/commit/view/settings/CommitWindow.java
package com.fobgochod.git.commit.view.settings;

import com.fobgochod.git.commit.ChangeType;
import com.fobgochod.git.commit.CommitMessage;
import com.fobgochod.git.commit.GitLogQuery;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.ComboBox;
import com.intellij.util.ui.FormBuilder;

import javax.swing.*;
import java.awt.*;
import java.io.File;

public class CommitWindow {

    private JPanel mainPanel;
    private JComboBox<ChangeType> changeType;
    private JComboBox<String> changeScope;
    private JTextField shortDescription = new JTextField();
    private JTextArea longDescription = new JTextArea();
    private JCheckBox wrapTextCheckBox = new JCheckBox("Wrap text at 72 characters?", true);
    private JTextArea breakingChanges = new JTextArea();
    private JTextField closedIssues = new JTextField();
    private JCheckBox skipCICheckBox = new JCheckBox("Skip CI?");

    public CommitWindow(Project project, CommitMessage commitMessage) {

        changeType = new ComboBox<>(new DefaultComboBoxModel<>(ChangeType.values()));
        changeScope = new ComboBox<>( );

        File workingDirectory = new File(project.getBasePath());
        GitLogQuery.Result result = new GitLogQuery(workingDirectory).execute();
        if (result.isSuccess()) {
            changeScope.addItem(""); // no value by default
            result.getScopes().forEach(changeScope::addItem);
        }

        if (commitMessage != null) {
            restoreValuesFromParsedCommitMessage(commitMessage);
        }

        longDescription.setPreferredSize(new Dimension(150, 100));
        longDescription.setLineWrap(true);
        breakingChanges.setPreferredSize(new Dimension(150, 50));
        breakingChanges.setLineWrap(true);

        mainPanel = FormBuilder.createFormBuilder()
                .addLabeledComponent(new JLabel("Type of change"), changeType)
                .addLabeledComponent(new JLabel("Scope of this change"), changeScope)
                .addLabeledComponent(new JLabel("Short description"), shortDescription)
                .addLabeledComponent(new JLabel("Long description"), longDescription)
                .addComponentToRightColumn(wrapTextCheckBox)
                .addLabeledComponent(new JLabel("Breaking changes"), breakingChanges)
                .addLabeledComponent(new JLabel("Closed issues"), closedIssues)
                .addComponentToRightColumn(skipCICheckBox)
                .getPanel();
    }

    public JPanel getMainPanel() {
        return mainPanel;
    }

    public CommitMessage getCommitMessage() {


        return new CommitMessage(
                getSelectedChangeType(),
                getSelectedChangeScope(),
                shortDescription.getText().trim(),
                longDescription.getText().trim(),
                breakingChanges.getText().trim(),
                closedIssues.getText().trim(),
                wrapTextCheckBox.isSelected(),
                skipCICheckBox.isSelected()
        );
    }

    private ChangeType getSelectedChangeType() {
        Object selectedItem = changeType.getSelectedItem();
        return (ChangeType) selectedItem;
    }

    private String getSelectedChangeScope() {
        Object selectedItem = changeScope.getSelectedItem();
        return (String) selectedItem;
    }

    private void restoreValuesFromParsedCommitMessage(CommitMessage commitMessage) {
        changeType.setSelectedItem(commitMessage.getChangeType());
        changeScope.setSelectedItem(commitMessage.getChangeScope());
        shortDescription.setText(commitMessage.getShortDescription());
        longDescription.setText(commitMessage.getLongDescription());
        breakingChanges.setText(commitMessage.getBreakingChanges());
        closedIssues.setText(commitMessage.getClosedIssues());
        skipCICheckBox.setSelected(commitMessage.isSkipCI());
    }
}

File: src/main/kotlin/com/fobgochod/git/commit/view/CommitWindow.kt
package com.fobgochod.git.commit.view

import com.fobgochod.git.commit.domain.CommitMessage
import com.fobgochod.git.commit.util.GitLogQuery
import com.fobgochod.git.commit.domain.TypeRow
import com.fobgochod.git.commit.settings.GitState
import com.fobgochod.git.commit.util.GitBundle
import com.intellij.icons.AllIcons
import com.intellij.openapi.options.ShowSettingsUtil
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ComboBox
import com.intellij.util.ui.FormBuilder
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.GridLayout
import java.awt.event.ItemEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.util.function.Consumer
import javax.swing.*

class CommitWindow(val project: Project?, private val oldMessage: CommitMessage) {

    private var state: GitState = GitState.getInstance()

    private val formBuilder = FormBuilder.createFormBuilder();

    private val changeTypePanel = JPanel()
    private val changeTypeGroup = ButtonGroup();
    private val changeType: JComboBox<TypeRow> = ComboBox()

    private val changeScopePanel = JPanel(BorderLayout())
    private val changeScope: JComboBox<String> = ComboBox()

    private val changeSubject = JTextField()
    private val changeBody = JTextArea(6, 0)
    private val wrapText = JCheckBox("Wrap text at 72 characters?", true)
    private val breakingChanges = JTextArea(3, 0)
    private val closedIssues = JTextField()

    private val bottomPanel: JPanel = JPanel(BorderLayout())
    private val skipCI = JCheckBox("Skip CI?")
    private val editSettings = JLabel(AllIcons.General.Settings)

    init {
        initView()
        initEvent()
        initData()
    }

    val root: JPanel get() = formBuilder.panel

    val commitMessage: CommitMessage
        get() = CommitMessage(
            selectedChangeType,
            selectedChangeScope,
            changeSubject.text.trim { it <= ' ' },
            changeBody.text.trim { it <= ' ' },
            wrapText.isSelected,
            breakingChanges.text.trim { it <= ' ' },
            closedIssues.text.trim { it <= ' ' },
            skipCI.isSelected
        )

    private fun initView() {
        var typeCount = state.commonCount;
        if (state.commonCount > state.typeRows.size) {
            typeCount = state.typeRows.size
        }

        changeTypePanel.layout = GridLayout(typeCount + 1, 1)
        for ((index, type) in state.typeRows.withIndex()) {
            if (index < state.commonCount) {
                val radioButton = JRadioButton(type.toString())
                changeTypeGroup.add(radioButton)
                changeTypePanel.add(radioButton)
            }
        }
        changeTypePanel.add(changeType);

        changeScope.isEditable = true
        changeScopePanel.add(changeScope, BorderLayout.CENTER)

        changeBody.lineWrap = true
        breakingChanges.lineWrap = true

        bottomPanel.add(skipCI, BorderLayout.CENTER)
        bottomPanel.add(editSettings, BorderLayout.EAST)

        formBuilder.panel.minimumSize = Dimension(600, 0)
        formBuilder.addLabeledComponent(JLabel("Type of change"), changeTypePanel)
            .addLabeledComponent(JLabel("Scope of change"), changeScopePanel)
            .addLabeledComponent(JLabel("Subject of change"), changeSubject)
            .addLabeledComponent(JLabel("Body of change"), changeBody)
            .addComponentToRightColumn(wrapText)
            .addLabeledComponent(JLabel("Breaking Changes"), breakingChanges)
            .addLabeledComponent(JLabel("Closed Issues"), closedIssues)
            .addComponentToRightColumn(bottomPanel)
    }

    private fun initEvent() {
        for (element in changeTypeGroup.elements) {
            element.addChangeListener {
                if (element.isSelected) {
                    changeType.selectedItem = GitState.getInstance().getTypeFromName(element.text)
                }
            }
        }

        changeType.addItemListener { e: ItemEvent ->
            val item: TypeRow = e.item as TypeRow
            if (state.typeRows.indexOf(item) > state.commonCount) {
                changeTypeGroup.clearSelection()
            } else {
                for (element in changeTypeGroup.elements) {
                    if (item == GitState.getInstance().getTypeFromName(element.text)) {
                        element.isSelected = true;
                    }
                }
            }
        }

        editSettings.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                ShowSettingsUtil.getInstance().showSettingsDialog(project, GitBundle.message("plugin.name"))
            }
        })
    }

    private fun initData() {
        // 恢复数据
        for (typeRow in GitState.getInstance().typeRows) {
            changeType.addItem(typeRow)
        }

        val result = GitLogQuery(project).execute()
        if (result.isSuccess()) {
            changeScope.addItem("") // no value by default
            result.getScopes().forEach(Consumer { item: String ->
                changeScope.addItem(
                    item
                )
            })
        }

        restoreFromParsedCommitMessage(oldMessage)
    }


    private val selectedChangeType: String
        get() {
            val selectedItem = changeType.selectedItem
            return (selectedItem as TypeRow).title
        }
    private val selectedChangeScope: String
        get() {
            val selectedItem = changeScope.selectedItem
            return selectedItem as String
        }

    private fun restoreFromParsedCommitMessage(commitMessage: CommitMessage) {
        changeType.selectedItem = commitMessage.changeType
        changeScope.selectedItem = commitMessage.changeScope
        changeSubject.text = commitMessage.changeSubject
        changeBody.text = commitMessage.changeBody
        wrapText.isSelected = commitMessage.wrapText
        breakingChanges.text = commitMessage.breakingChanges
        closedIssues.text = commitMessage.closedIssues
        skipCI.isSelected = commitMessage.skipCI
    }
}
